src/main/java/carpet/patches/EntityPlayerMPFake.java

```diff
@@ -11,45 +11,87 @@
 import net.minecraft.entity.damage.DamageSource;
 import net.minecraft.entity.player.HungerManager;
 import net.minecraft.network.NetworkSide;
+import net.minecraft.server.PlayerManager;
 import net.minecraft.text.Text;
 import net.minecraft.text.TranslatableText;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.ServerTask;
 import net.minecraft.server.network.ServerPlayerEntity;
 import net.minecraft.server.network.ServerPlayerInteractionManager;
 import net.minecraft.server.world.ServerWorld;
-import net.minecraft.util.UserCache;
 import net.minecraft.util.registry.RegistryKey;
 import net.minecraft.world.GameMode;
 import net.minecraft.world.World;
-import net.minecraft.world.dimension.DimensionType;
 import carpet.fakes.ServerPlayerEntityInterface;
 import carpet.utils.Messenger;
 
+import java.util.Objects;
+
 @SuppressWarnings("EntityConstructor")
 public class EntityPlayerMPFake extends ServerPlayerEntity
 {
     public Runnable fixStartingPosition = () -> {};
     public boolean isAShadow;
 
-    public static EntityPlayerMPFake createFake(String username, MinecraftServer server, double d0, double d1, double d2, double yaw, double pitch, RegistryKey<World> dimensionId, GameMode gamemode)
-    {
-        //prolly half of that crap is not necessary, but it works
-        ServerWorld worldIn = server.getWorld(dimensionId);
-        ServerPlayerInteractionManager interactionManagerIn = new ServerPlayerInteractionManager(worldIn);
-        UserCache.setUseRemote(false);
-        GameProfile gameprofile;
-        try {
-            gameprofile = server.getUserCache().findByName(username);
+    public static EntityPlayerMPFake createFake(String playerName, MinecraftServer server, double d0, double d1, double d2, double yaw, double pitch, RegistryKey<World> dimensionId, GameMode gamemode, boolean isSpawnerPrivileged, boolean fallBackToSpawningOfflinePlayer
+    ) throws FakePlayerSpawnException {
+        Objects.requireNonNull(playerName);
+        Objects.requireNonNull(server);
+        Objects.requireNonNull(dimensionId);
+        Objects.requireNonNull(gamemode);
+
+        PlayerManager playerManager = server.getPlayerManager();
+
+        // if this player already logged in
+        if (playerManager.getPlayer(playerName) != null)
+        {
+            throw new FakePlayerSpawnException("r Player ", "rb " + playerName, "r  is already logged on");
+        }
+
+        // check player name length
+        // player with a name longer than 16 bytes will kick out all players
```

Comment:
not true - works in SP up to 40 chars

Reply:
Well, I saw that another PR is focusing on this problem, I will remove it from this PR.